expr/functions/exponentialMovingAverage/function.go

```diff
@@ -0,0 +1,138 @@
+package exponentialMovingAverage
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/go-graphite/carbonapi/expr/consolidations"
+	"github.com/go-graphite/carbonapi/expr/helper"
+	"github.com/go-graphite/carbonapi/expr/interfaces"
+	"github.com/go-graphite/carbonapi/expr/types"
+	"github.com/go-graphite/carbonapi/pkg/parser"
+)
+
+type exponentialMovingAverage struct {
+	interfaces.FunctionBase
+}
+
+func GetOrder() interfaces.Order {
+	return interfaces.Any
+}
+
+func New(configFile string) []interfaces.FunctionMetadata {
+	res := make([]interfaces.FunctionMetadata, 0)
+	f := &exponentialMovingAverage{}
+	functions := []string{"exponentialMovingAverage"}
+	for _, n := range functions {
+		res = append(res, interfaces.FunctionMetadata{Name: n, F: f})
+	}
+	return res
+}
+
+func (f *exponentialMovingAverage) Do(ctx context.Context, e parser.Expr, from, until int64, values map[parser.MetricRequest][]*types.MetricData) ([]*types.MetricData, error) {
+	var n int
+	var err error
+
+	var argstr string
+
+	if len(e.Args()) < 2 {
+		return nil, parser.ErrMissingArgument
+	}
+
+	switch e.Args()[1].Type() {
+	case parser.EtConst:
+		// In this case, zipper does not request additional retrospective points,
+		// and leading `n` values, that used to calculate window, become NaN
+		n, err = e.GetIntArg(1)
+		argstr = strconv.Itoa(n)
+	case parser.EtString:
+		var n32 int32
+		n32, err = e.GetIntervalArg(1, 1)
+		if err != nil {
+			return nil, err
+		}
+		argstr = fmt.Sprintf("%q", e.Args()[1].StringValue())
+		n = int(n32)
+	default:
+		err = parser.ErrBadType
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*types.MetricData
+	windowSize := n
+
+	start := from
+
+	arg, err := helper.GetSeriesArg(ctx, e.Args()[0], start, until, values)
+	if err != nil {
+		return nil, err
+	}
+	if len(arg) == 0 {
+		return results, nil
+	}
+
+	constant := float64(2 / (float64(windowSize) + 1))
+
+	for _, a := range arg {
+		r := a.CopyLink()
+		r.Name = fmt.Sprintf("%s(%s,%s)", e.Target(), a.Name, argstr)
```

Comment:
Also mayby without dynamic Sprintf (which need scan format string and uses high count interfaces boxing/unboxing) ?